File: src/main/org/nlogo/prim/etc/_subject.java
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.api.Agent;
import org.nlogo.api.PerspectiveJ;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Reporter;

public final strictfp class _subject
    extends Reporter {
  @Override
  public Object report(final org.nlogo.nvm.Context context) {
    if (world.observer().perspective() == PerspectiveJ.OBSERVE()) {
      return org.nlogo.api.Nobody$.MODULE$;
    }
    Agent subject = world.observer().targetAgent();
    // not actually sure if the null check here is necessary - ST 6/28/05
    if (subject == null || subject.id() == -1) {
      return org.nlogo.api.Nobody$.MODULE$;
    }
    return subject;
  }

  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax(Syntax.AgentType());
  }
}

File: src/main/org/nlogo/prim/etc/_subject.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.api.{ Nobody, Perspective, Syntax }
import org.nlogo.nvm.{ Context, Reporter }

class _subject extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(Syntax.AgentType)
  override def report(context: Context) =
    if (world.observer.perspective == Perspective.Observe)
      Nobody
    else {
      val subject = world.observer.targetAgent
      // not actually sure if the null check here is necessary - ST 6/28/05
      if (subject == null || subject.id == -1)
        Nobody
      else
        subject
    }
}
